src/btree/bt_cursor.c

```diff
@@ -481,19 +482,32 @@ __wt_btcur_search_uncommitted(WT_CURSOR_BTREE *cbt, WT_UPDATE **updp)
      * itself. But, we cannot be sure of finding one, as pre processing of this prepared transaction
      * updates could have happened as part of resolving earlier transaction operations.
      */
-    if (cbt->compare != 0)
-        return (0);
+    if (cbt->compare == 0) {
+        /*
+         * Get the uncommitted update from the cursor. For column store there will be always a
+         * insert structure for updates irrespective of fixed length or variable length.
+         */
+        if (cbt->ins != NULL)
+            upd = cbt->ins->upd;
+        else if (cbt->btree->type == BTREE_ROW)
+            if (cbt->ref->page->modify)
+                upd = cbt->ref->page->modify->mod_row_update[cbt->slot];
+    }
 
     /*
-     * Get the uncommitted update from the cursor. For column store there will be always a insert
-     * structure for updates irrespective of fixed length or variable length.
+     * In the case of prepare, there could be "uncommitted" updates in the lookaside that are newer
+     * than those in the update list.
      */
-    if (cbt->ins != NULL)
-        upd = cbt->ins->upd;
-    else if (cbt->btree->type == BTREE_ROW) {
-        WT_ASSERT(session, cbt->btree->type == BTREE_ROW && cbt->ref->page->modify != NULL &&
-            cbt->ref->page->modify->mod_row_update != NULL);
-        upd = cbt->ref->page->modify->mod_row_update[cbt->slot];
+    if (__wt_page_las_active(session, cbt->ref) && cbt->ref->page_las->has_las &&
+      cbt->ref->page_las->has_prepares) {
+        ret = __wt_find_lookaside_upd(session, cbt, &tmp_upd, true);
+        WT_RET_NOTFOUND_OK(ret);
+        if (ret == 0) {
+            if (upd == NULL || upd->start_ts < tmp_upd->start_ts)
+                upd = tmp_upd;
```

Comment:
Are we assuming that the update we found is definitely going to be in the prepared state? I think it is a fair assumption as we "known" that for this key we have a prepared update and if the update we find in lookaside is the newest it must be in the prepared state. Let me know if my understanding is wrong.

I think we need to assert our assumption though: Once we have found a newer update in the lookaside, we should assert that this update is in the prepared state.

Reply:
fd82c8e

Your understanding is correct. There is a comment indicating above that it's specifically for the prepare case. I've added an assertion to check that if we end up using the lookaside update that it truly is a prepare.